### PR TITLE
Cloud Build

### PR DESCRIPTION
**Pull request by AndrewDryga**
A sample of cloud build config that we might use.

This speeds up Elixir build substantially by leveraging the layer cache stored in the Google Artifact Registry.

The extra trick which I should we should keep no matter how we build our images: is compiling Elixir in a separate `compiler` step, which will allow us to greatly reduce concurrent compile time for `api`/`web` images because 80% of releasing them is fetching deps and compiling the source.

Also, we can bypass issues with GitHub Actions Cache size by using Google Artifact Registry even if we keep using GH Actions.

Compiling the source without cache (which is shared) takes 4 minutes.
On top of that releasing api/web takes ~32 seconds.
See https://console.cloud.google.com/cloud-build/builds;region=global/8d21c772-b6b7-4308-a580-
d201db517765;step=3?authuser=1&project=firezone-staging.

With a cached source, everything takes 56 seconds:
https://console.cloud.google.com/cloud-build/builds;region=global/d828775b-40b4-47cb-b10f-a6517b72ca4d;step=4?authuser=1&project=firezone-staging

With a change in one of the apps, everything takes 1 minute 43 seconds:
https://console.cloud.google.com/cloud-build/builds;region=global/a66adaff-e7f2-4dff-bf96-fafdcfc0ed5d?authuser=1&project=firezone-staging

To try it yourself run:
```bash
gcloud builds submit --config cloudbuild.yaml --substitutions='_REPOSITORY=firezone,_REGISTRY=us-east1-docker.pkg.dev,_API_TAG=local,_WEB_TAG=local' --project "firezone-staging" --timeout "30m"
```

Things we can improve: fetch and build assets before doing so for Elixir code which changes way more often.

### CODE REVIEW

**Comment by AndrewDryga** [comment]
Added rust part with changes pulled from #2233, clean rust build for *relay AND gateway* take 5:46 instead of 15 minutes just for the relay:
```
PUSH
DONE
---------------------------------------------------------------------------------------------------------------------------------------------------------
ID                                    CREATE_TIME                DURATION  SOURCE                                                                                          IMAGES  STATUS
78aeed73-e196-4b67-8360-d689948ab720  2023-10-04T02:33:15+00:00  5M48S     gs://firezone-staging_cloudbuild/source/1696386756.305391-688de90ebb7549659c1c32548772caa1.tgz  -       SUCCESS
```
see https://console.cloud.google.com/cloud-build/builds;region=global/78aeed73-e196-4b67-8360-d689948ab720?authuser=1&project=firezone-staging

**Comment by AndrewDryga** [comment]
Fully cached relay/gateway build takes just 7 seconds: https://console.cloud.google.com/cloud-build/builds;region=global/095e3771-070f-4b9b-a1b9-25a27e1d787e?authuser=1&project=firezone-staging

After I changed one line in the relay source code the deps are cached (hooray!) both relay and gateway are recompiled (unfortunately), complete build takes under 5:48, with ~1:30 spent on gzipping and uploading the rust build cache: https://console.cloud.google.com/cloud-build/builds;region=global/2f285103-a3ec-4f8b-bad5-9b5abc030b71;step=6?authuser=1&project=firezone-staging. I think the issue here is the volume of the cache, on a very fast instance it takes about the same time to download the cache as to rebuild it.

6 minutes for all containers is acceptable, it's as much as our fastest container with GitHub actions.

**Comment by AndrewDryga** [comment]
> I think the issue here is the volume of the cache, on a very fast instance it takes about the same time to download the cache as to rebuild it.

To prove the point here is a build with all rust caches disabled: https://console.cloud.google.com/cloud-build/builds;region=global/b4bc0a1c-aa02-4e46-a9f0-dc4ff83105f9?authuser=1&project=firezone-staging

*Takes 1 minute less!*

**Comment by AndrewDryga** [comment]
@jamilbk trying to cleanup the CI here is what I came up with: https://github.com/firezone/firezone/actions/runs/6401336878 This feels way cleaner to have a workflow file per language and then use them only as part of bigger workflows, allowing more reuse. 

I removed conditions for now and we build everything, but I will bring them back alter. This is something we need to discuss: do we disable merge queues and those "fake test pass" or we run all tests all the time making everything much slower.

cc @conectado @thomaseizinger

**Review by jamilbk — COMMENTED** [review]
I'm seeing similar speeds with the GitHub runners -- now that the CI caching is fixed, API/Web builds take about 30s as well, under a minute for the full thing:

https://github.com/firezone/firezone/actions/runs/6402910501/job/17380466144

I appreciate the work that went into this, and I think we definitely might use cloud build in the future for more building our production containers in CD, but I'm not sure it's wise a good idea to introduce another build system into our stack right now.

Also, I think you need a GCloud account in our organization to view these logs?

GCloud builders could be used for our prod images (api/web/relay/gateway) perhaps later on.

File: .github/workflows/ci.yml
Comment (jamilbk) [inline .github/workflows/ci.yml:1]: Thanks for cleaning this up! I was just about to do something similar in my PR 👍.

**Comment by jamilbk** [comment]
> @jamilbk trying to cleanup the CI here is what I came up with: https://github.com/firezone/firezone/actions/runs/6401336878 This feels way cleaner to have a workflow file per language and then use them only as part of bigger workflows, allowing more reuse.
> 
> I removed conditions for now and we build everything, but I will bring them back alter. This is something we need to discuss: do we disable merge queues and those "fake test pass" or we run all tests all the time making everything much slower.
> 
> cc @conectado @thomaseizinger

@AndrewDryga The merge queue had nothing to do with those fake `pass-checks` workflows. Those were required because we're using [path-based conditions along with "Require status checks to pass"](https://github.com/orgs/community/discussions/44490) -- some workflows would never be triggered if their paths hadn't changed, making the PR unmergeable.

Removing those path-based conditions means all our workflows run on each PR, which I don't think is necessarily a bad thing _if our minutes are free_. Running the Gcloud builder on _each PR_ could get expensive fast though, hence one of the reasons for wanting to keep this in GH for now.

File: .github/workflows/ci.yml
Comment (AndrewDryga) [reply .github/workflows/ci.yml:1]: Yeah, that's the part I want you take from this PR the most :).

**Comment by AndrewDryga** [comment]
> I'm seeing similar speeds with the GitHub runners -- now that the CI caching is fixed, API/Web builds take about 30s as well, under a minute for the full thing:

That's great, I think then we should keep building on GitHub Actions, especially because we can use a larger runner anytime we want.

> I appreciate the work that went into this, and I think we definitely might use cloud build in the future for more building our production containers in CD, but I'm not sure it's wise a good idea to introduce another build system into our stack right now.

I agree, if it doesn't save at least 50% of our time it's not worth to maintain it.

> Also, I think you need a GCloud account in our organization to view these logs?

Not at all, we can use a custom bucket for logs and then they will be streamed to GitHub console output.

> GCloud builders could be used for our prod images (api/web/relay/gateway) perhaps later on.

Let's see how long it will take to do a multi-arch build for our rust targets. I feel like we will have to do bigger instances to get any meaningful time there, but with GitHub we can also try to build them on native platforms. The only massive disadvantage I see here for now is the very limited cache size, but we can totally link our GitHub Actions caches to the Google Artifact Registry.

**Comment by AndrewDryga** [comment]
> Removing those path-based conditions means all our workflows run on each PR, which I don't think is necessarily a bad thing if our minutes are free. Running the Gcloud builder on each PR could get expensive fast though, hence one of the reasons for wanting to keep this in GH for now.

Yeah, that's what I meant. Only I worry about developer minutes more than runner minutes, because having somebody to wait for 10 minutes more to run all the tests is not a great feeling, especially if you need to do it at least 3 times to deploy a hotfix to production. I would always spend any money on fast CI not having developers waiting an extra minute.

**Comment by jamilbk** [comment]
@AndrewDryga Yeah, I'm with you. We can use dedicated GH runners for certain jobs. Looks like they now support M1 macs too so we can build natively on `arm64`:

https://github.blog/2023-10-02-introducing-the-new-apple-silicon-powered-m1-macos-larger-runner-for-github-actions/

**Comment by thomaseizinger** [comment]
> > GCloud builders could be used for our prod images (api/web/relay/gateway) perhaps later on.
> 
> Let's see how long it will take to do a multi-arch build for our rust targets. I feel like we will have to do bigger instances to get any meaningful time there, but with GitHub we can also try to build them on native platforms. The only massive disadvantage I see here for now is the very limited cache size, but we can totally link our GitHub Actions caches to the Google Artifact Registry.

Buildkit also has an S3 backend that you can push caches to: https://docs.docker.com/build/cache/backends/s3/
We use this for our interoperability tests in libp2p: https://github.com/libp2p/test-plans/blob/master/multidim-interop/dockerBuildWrapper.sh

I don't know whether that works with the github action but I'd be surprised if that one does any additional validation on which cache options you are allowed to pass. According to that page, using GitHub actions cache is also experimental.

**Comment by thomaseizinger** [comment]
> I don't know whether that works with the github action but I'd be surprised if that one does any additional validation on which cache options you are allowed to pass. According to that page, using GitHub actions cache is also experimental.

Worst case, we can just ditch the action and invoke buildkit ourselves.

**Pull request closed without merging**